controllers/buildpipeline/buildpipeline_adapter.go

```diff
@@ -129,6 +132,30 @@ func (a *Adapter) EnsureSnapshotExists() (controller.OperationResult, error) {
 		return controller.RequeueWithError(err)
 	}
 
+	return a.removeFinalizerAndContinueProcessing()
+}
+
+func (a *Adapter) EnsurePipelineIsFinalized() (controller.OperationResult, error) {
+	if h.HasPipelineRunFinished(a.pipelineRun) || controllerutil.ContainsFinalizer(a.pipelineRun, h.BuildPipelineRunFinalizer) {
+		return controller.ContinueProcessing()
+	}
+
+	err := retry.OnError(retry.DefaultRetry, func(_ error) bool { return true }, func() error {
+		patch := client.MergeFrom(a.pipelineRun.DeepCopy())
+		if !controllerutil.AddFinalizer(a.pipelineRun, h.BuildPipelineRunFinalizer) {
```

Comment:
I still think we should use [RemoveFinalizerFromPipelineRun](https://github.com/redhat-appstudio/integration-service/blob/main/helpers/integration.go#L411-L425) here instead, especially since the current code has potential for infinite reconciliations if the finalizer is already set.

Reply:
Why would we use `RemoveFinalizerFromPipelineRun()` to add a finalizer?

Reply:
What I can do is verify that the finalizer doesn't already exist.  I can also create a function in helpers called `AddFinalizerToPipelineRun()` that handles all of this.

Reply:
Sorry, my bad, I thought we had an AddFinalizer function already and didn't check what I was c/p-ing. You're right, creating an `AddFinalizerToPipelineRun()` is the way to go.